lib/iris/experimental/um.py

```diff
@@ -508,37 +550,39 @@ def constants(name, dtype):
-            if lookup[Field.LBNREC_OFFSET, 0] == 0:
+            is_model_dump = lookup[Field.LBNREC_OFFSET, 0] == 0
+            if is_model_dump:
                 # A model dump has no direct addressing - only relative,
                 # so we need to update the offset as we create each
                 # Field.
                 running_offset = ((self.fixed_length_header.data_start - 1) *
                                   word_size)
-                for raw_headers in lookup.T:
-                    if raw_headers[0] == -99:
-                        data_provider = None
-                    else:
+
+            for raw_headers in lookup.T:
+                ints = raw_headers[:_NUM_FIELD_INTS]
+                reals = raw_headers[_NUM_FIELD_INTS:].view(real_dtype)
+                field_class = _FIELD_CLASSES.get(ints[Field.LBREL_OFFSET],
+                                                 Field)
+                if raw_headers[0] == -99:
+                    data_provider = None
+                else:
+                    if is_model_dump:
                         offset = running_offset
-                        data_provider = data_class(source, offset, word_size)
-                    klass = _FIELD_CLASSES.get(raw_headers[Field.LBREL_OFFSET],
-                                               Field)
-                    ints = raw_headers[:_NUM_FIELD_INTS]
-                    reals = raw_headers[_NUM_FIELD_INTS:].view(real_dtype)
-                    fields.append(klass(ints, reals, data_provider))
-                    running_offset += (raw_headers[Field.LBLREC_OFFSET] *
-                                       word_size)
-            else:
-                for raw_headers in lookup.T:
-                    if raw_headers[0] == -99:
-                        data_provider = None
                     else:
                         offset = raw_headers[Field.LBEGIN_OFFSET] * word_size
-                        data_provider = data_class(source, offset, word_size)
-                    klass = _FIELD_CLASSES.get(raw_headers[Field.LBREL_OFFSET],
-                                               Field)
-                    ints = raw_headers[:_NUM_FIELD_INTS]
-                    reals = raw_headers[_NUM_FIELD_INTS:].view(real_dtype)
-                    fields.append(klass(ints, reals, data_provider))
+                    # Make a *copy* of field lookup data, as it was in the
+                    # untouched original file, as a context for data loading.
+                    # (N.B. most importantly, includes the original LBPACK)
+                    lookup_reference = field_class(ints[:], reals[:], None)
+                    # Make a "provider" that can fetch the data on request.
+                    data_provider = data_class(source, filename,
+                                               lookup_reference,
+                                               offset, word_size)
+                field = field_class(ints, reals, data_provider)
```

Comment:
> is this a stepping stone or a sign of things to come?

So if I can summarise... there are no plans to move away from the Field-Provider-Field pattern, but there are no fundamental obstacles to doing so - it would just need a simple-enough, clean-enough implementation.

Thanks.

Move along now ... nothing to see here. :see_no_evil: